baseapp/abci_utils.go

```diff
@@ -207,12 +218,31 @@ func (h DefaultProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHand
 					panic(err)
 				}
 			} else {
+				var txGasLimit uint64
 				txSize := int64(len(bz))
-				if totalTxBytes += txSize; totalTxBytes <= req.MaxTxBytes {
-					selectedTxs = append(selectedTxs, bz)
-				} else {
-					// We've reached capacity per req.MaxTxBytes so we cannot select any
-					// more transactions.
+
+				gasTx, ok := memTx.(GasTx)
+				if ok {
+					txGasLimit = gasTx.GetGas()
+				}
+
+				// only add the transaction to the proposal if we have enough capacity
+				if (txSize + totalTxBytes) < req.MaxTxBytes {
+					// If there is a max block gas limit, add the tx only if the limit has
+					// not been met.
+					if maxBlockGas > 0 && (txGasLimit+totalTxGas) <= uint64(maxBlockGas) {
+						totalTxGas += txGasLimit
+						totalTxBytes += txSize
+						selectedTxs = append(selectedTxs, bz)
+					} else {
+						totalTxBytes += txSize
+						selectedTxs = append(selectedTxs, bz)
+					}
+				}
+
+				// Check if we've reached capacity. If so, we cannot select any more
+				// transactions.
+				if totalTxBytes >= req.MaxTxBytes || totalTxGas >= uint64(maxBlockGas) {
```

Comment:
I think the issue is here. If maxBlockGas == 0 then we should ignore totalTxGas and keep adding txs if `totalTxBytes >= req.MaxTxBytes` hasn't been met

Reply:
And the other issue is that one of the tests is trying to make txs with a huge gas limit which is now prohibited. Would this be a breaking change somehow?